working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -41,126 +42,253 @@ Since a mixin application class cannot contain static members, there is no risk
 ```
 
-This is not always as useful as it seems because it doesn't work when the superclass constructor has optional positional parameters. You cannot usefully extend `DateTime` the same way because the only constructor which doesn't have optional positional parameters is `DateTime.now`.
+This will require some constraints on the superclass constructors, though. For example, we can't have both optional positional parameters and named parameters on the same class.
 
-It is still a very useful feature when you simply add methods to a class:
+It is still a very useful feature when you simply add methods to a class, and no new fields:
 
 ```dart
 class MyRectangle extends Rectangle<int> {
-  MyRectangle affineTransform(int a, int b, int c, int d) => 
-      MyRectangle(a * left + b * top, c * left + d * top, 
+  MyRectangle affineTransform(int a, int b, int c, int d) =>
+      MyRectangle(a * left + b * top, c * left + d * top,
                   a * right + b * bottom, c * right + d * bottom);
 }
 ```
 
-Here you implicitly "inherit" all constructors from the superclass without having to rewrite them.
+Here you could implicitly "inherit" all constructors from the superclass without having to rewrite them.
+
+## Proposal
+
+We will introduce *initializing* and *forwarding* constructors, and have *default constructors* be both.
+
+### Initializing Constructor
+
+An initializing constructor for a class *C* is a generative constructor with *implicitly* introduced named initializing formals without default values for each instance variable declared in *C* which is not declared `late` and which does not have an initializer expression. The named parameter is required if the instance variable's type is potentially non-nullable. In NNBD-legacy code, the parameter is required if the instance variable is final.
+
+A constructor can be made initializing by writing `default` as a modifier before the constructor, after any `const` modifier.
+
+Such a constructor must not declare any optional positional parameters. It may declare other parameters, and then an initializing formal is not introduced for an instance variable when the constructor declares another parameter with the same name. This allows users to explicitly initialize some fields and still have default initialization for the remaining fields.
+
+#### Example
+
+The class `Person` above could be given an initializing constructor by declaring a constructor of the form:
+
+```dart
+  default Person();
+```
+
+This would expand to the constructor:
 
-## Notes
+```dart
+  Person({required this.firstName, required this.lastName, required this.age}) : super();
+```
+
+An initializing constructor of the form:
+
+```dart
+  default Person(this.lastName);
+```
+
+would expand to;
+
+```dart
+  Person(this.lastName, {required this.firstName, required this.age}) : super()
+
+```
+
+because the explicitly supplied `lastName` parameter takes precedence over inserting an initializing formal parameter for that field.
```

Comment:
Syntax bikeshed: Now that the syntax is really sugar for *part of a parameter list* and not *an entire constructor*, consider putting the syntax inside the parameter list:

```dart
  Person(this.lastName, default);
```

This would offer some interesting affordances:

```dart
  Person(this.lastName, default); // The field initializers are positional and required.
  Person(this.lastName, [default]); // The field initializers are positional and optional.
  Person(this.lastName, {default}); // The field initializers are named and optional.
  Person(this.lastName, {required default}); // The field initializers are named and required.

  Person(default, this.lastName); // Default ones come first.
```

Reply:
If we went this route, what about using `this` instead of `default`?  In the initializer list, you could then also allow `this.firstName` to refer to the implicit `firstName` parameter.

Reply:
Putting the field initializer marker into the parameter list as *positional* parameters means that we get arbitrary ordering (source order most likely, because that's the only thing the user can actually control). I'd prefer to avoid that if possible.
That's why I put `default` outside, because it definitely will generate named parameters, and I didn't want to put two markers into the parameter list (the forwarding marker and the initializing marker). It gets too crowded.

If we did put it inside the parameter list, then `this` is actually a good marker, and it compares well with `super` for the forwarding parameters, and to the `this.name` for a specific field initializer. 
I'm just worried that if we *only* allow initializing parameters as named, then people will forget the `{...}` braces and just write `Foo(this)` instead of `Foo({this})`, and if we do allow them outside, users will have to order their fields for the constructor, not for what reads better.
(That said, I'm sure a lot of users will love it anyway, so maybe I'm being too careful).

Reply:
So, brainstorm on allowing `this` as the "intialize all" marker anywhere in the parameter list.

For each class, there are a given number of auto-initializable fields (non-late, non-initializer-expression) fields. For named initialization, they must also be non-private. We probably want to omit private fields everywhere, and require you to do those manually, because otherwise we leak the private name (and leave you with no way to *avoid* the auto-initialization of a field).
Some of those initializers will have to be required

So, if you write `this` in the positional parameter part, then they will all be positional (even the boolean ones, which some people will scoff at). 

If it's in the optional positional parameter part `[this]`, then they must all be optional. If some are necessarily required, then that's an error.

If it's in the required positional parameter part, with any required parameter after it, then they must all be required. That's acceptable.

If it's at the *end* of the required positional parameter part, then we can make the tail of the parameters optional. Any required initializer makes all prior (source order) field initializers also required, but if the final ones are nullable, they can become optional.

It feels fragile.

If the `this` is in the named parameter part, then it works as now. Some are optional, those who must be are required.
If it's modified by `required`, then they're all required.

I'm not sure this buys use enough to be worth the complexity and general footgunnery.